socket.go

```diff
@@ -202,27 +185,23 @@ func shouldAcceptNewConnectionHandler(listenerPtr, connPtr, devicePtr unsafe.Poi
 //
 // see: https://developer.apple.com/documentation/virtualization/vzvirtiosocketconnection?language=objc
 type VirtioSocketConnection struct {
-	sourcePort      uint32
-	destinationPort uint32
-	fileDescriptor  uintptr
-	file            *os.File
-	laddr           net.Addr // local
-	raddr           net.Addr // remote
+	net.Conn
+	laddr *Addr // local
+	raddr *Addr // remote
 }
 
 var _ net.Conn = (*VirtioSocketConnection)(nil)
 
-func newVirtioSocketConnection(ptr unsafe.Pointer) *VirtioSocketConnection {
+func newVirtioSocketConnection(ptr unsafe.Pointer) (*VirtioSocketConnection, error) {
 	vzVirtioSocketConnection := C.convertVZVirtioSocketConnection2Flat(ptr)
-	err := unix.SetNonblock(int(vzVirtioSocketConnection.fileDescriptor), true)
+	file := os.NewFile((uintptr)(vzVirtioSocketConnection.fileDescriptor), "")
+	defer file.Close()
```

Comment:
This line is unnecessary because of closed the file descriptor in the virtualization framework.
```suggestion
```

Reply:
If we don't explicitly close the file, go garbage collector may do it for us: https://pkg.go.dev/os#File.Fd
>  If f is garbage collected, a finalizer may close the file descriptor, making it invalid; 

This is problematic if the following sequence happens:
- `file := os.NewFile((uintptr)(vzVirtioSocketConnection.fileDescriptor), "")`
- virtualization framework calls `vzVirtioSocketConnection.fileDescriptor`
- another parts of the code needs a file descriptor, the OS/system library decides to use the same fd as `vzVirtioSocketConnection.fileDescriptor`
- the go garbage collector closes the file descriptor used by `file`, which is no longer the expected one and will cause unexpected issues

Reply:
I'm not sure why needs to call close method from your comment

Reply:
See https://gist.githubusercontent.com/cfergeau/46a4a4fda70a83dbe2a9bf519cbabd28/raw/530c7d9a97e7a823fb324552bda470ad11fbbab9/fd.go 
This shows the race which can happen when you call `os.NewFile()` on a file descriptor, and then other code closes the file descriptor.

By calling Close() in `newVirtioSocketConnection`, we know we are closing the right file descriptor. The garbage collector won't call `Close()` on it later.

Reply:
If `vzVirtioSocketConnection.fileDescriptor` is owned by the virtualization framework, the code below would even be more correct. Since `net.FileConn()` will also `Dup()` the file descriptor, this feels a bit overkill :-/

```diff
diff --git a/socket.go b/socket.go
index 173cae6..5770618 100644
--- a/socket.go
+++ b/socket.go
@@ -12,6 +12,7 @@ import (
        "os"
        "runtime"
        "runtime/cgo"
+       "syscall"
        "unsafe"
 
        "golang.org/x/sys/unix"
@@ -192,7 +193,11 @@ type VirtioSocketConnection struct {
 
 func newVirtioSocketConnection(ptr unsafe.Pointer) (*VirtioSocketConnection, error) {
        vzVirtioSocketConnection := C.convertVZVirtioSocketConnection2Flat(ptr)
-       file := os.NewFile((uintptr)(vzVirtioSocketConnection.fileDescriptor), "")
+       fd, err := syscall.Dup(int(vzVirtioSocketConnection.fileDescriptor))
+       if err != nil {
+               return nil, err
+       }
+       file := os.NewFile(uintptr(fd), "")
        defer file.Close()
        rawConn, err := net.FileConn(file)
        if err != nil {
```

Reply:
@cfergeau Yes, it is because only Go world.

I'm thinking your code behavior is

1. Close file descriptor by this line in Go
2. Virtualization framework will be tried to close the original file descriptor but already closed in Go world.

So I believe the original file descriptor should be closed in the framework (in Objective-C) instead of Go side.

I agree to call `Close` method for duplicated file descriptor in Go

Reply:
I didn't see this comment https://github.com/Code-Hex/vz/pull/63/files#r993303806 and I disagree this comment changes...

Reply:
net.FileConn will dup file descriptor. https://pkg.go.dev/net#FileConn
> FileConn returns a copy of the network connection corresponding to the open file f.

Reply:
> So I believe the original file descriptor should be closed in the framework (in Objective-C) instead of Go side.

The problem is that when `os.NewFile(fd)` is used, go will always call `Close(fd)`. If our code don't call Close(), then the go runtime/go garbage collector will call `Close()` automatically. 

> But I agree to call Close method for duplicated file descriptor in Go

Are you referring to `net.FileConn(file)` ? Or to another file descriptor?

Reply:
> If our code don't call Close(), then the go runtime/go garbage collector will call Close() automatically.

This is done here: https://github.com/golang/go/blob/4bcf94b0232db65ed5df47e0127cdbc8866aec64/src/os/file_unix.go#L196

Reply:
@cfergeau I know the Go call Close method in finalizer. Bu the timing is different with using defer.

Yes. about net.FileConn which dup from original file descriptor.

> Are you referring to net.FileConn(file) ? Or to another file descriptor?

Reply:
I don't mind either now.  Think about it after we have a problem.